Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp

```diff
@@ -0,0 +1,216 @@
+  IniFile game_ini_local;
+  game_ini_local.Load(File::GetUserPath(D_GAMESETTINGS_IDX) + m_game_id + ".ini");
+
+  m_gecko_codes = Gecko::LoadCodes(game_ini_default, game_ini_local);
+
+  UpdateList();
+}
+
+void GeckoCodeWidget::CreateWidgets()
+{
+  auto* layout = new QVBoxLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_name_label = new QLabel;
+  m_creator_label = new QLabel;
+  m_code_description = new QTextEdit;
+  m_code_view = new QTextEdit;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  QFont monospace(QStringLiteral("Monospace"));
+
+  if (monospace.family() != QStringLiteral("Monospace"))
+    monospace = QFontDatabase::systemFont(QFontDatabase::FixedFont);
+
+  m_code_description->setFont(monospace);
+  m_code_view->setFont(monospace);
+
+  m_code_description->setReadOnly(true);
+  m_code_view->setReadOnly(true);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list);
+  layout->setStretch(1, 3);
+
+  auto* info_layout = new QFormLayout;
+
+  info_layout->addRow(tr("Name:"), m_name_label);
+  info_layout->addRow(tr("Creator:"), m_creator_label);
+  info_layout->addRow(tr("Description:"), static_cast<QWidget*>(nullptr));
+
+  info_layout->setFormAlignment(Qt::AlignLeft | Qt::AlignTop);
+
+  for (QLabel* label : {m_name_label, m_creator_label})
+  {
+    label->setTextInteractionFlags(Qt::TextSelectableByMouse);
+    label->setCursor(Qt::IBeamCursor);
+  }
+
+  layout->addLayout(info_layout);
+  layout->addWidget(m_code_description);
+  layout->addWidget(m_code_view);
+  layout->addWidget(m_download_codes, 0, Qt::AlignRight);
+
+  m_download_codes->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Maximum);
+  m_code_description->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+  m_code_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+
+  setLayout(layout);
+}
+
+void GeckoCodeWidget::ConnectWidgets()
+{
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this,
+          &GeckoCodeWidget::OnSelectionChanged);
+  connect(m_code_list, &QListWidget::itemChanged, this, &GeckoCodeWidget::OnItemChanged);
+
+  connect(m_download_codes, &QPushButton::pressed, this, &GeckoCodeWidget::DownloadCodes);
+
+  connect(m_warning, &CheatWarningWidget::OpenCheatEnableSettings, this,
+          &GeckoCodeWidget::OpenGeneralSettings);
+
+  connect(this, &GeckoCodeWidget::EmulationStarted, m_warning,
+          &CheatWarningWidget::EmulationStarted);
+  connect(this, &GeckoCodeWidget::EmulationStopped, m_warning,
+          &CheatWarningWidget::EmulationStopped);
+}
+
+void GeckoCodeWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto selected = items[0];
+
+  const auto& code = m_gecko_codes[m_code_list->row(selected)];
+
+  QString notes;
+
+  for (const std::string& line : code.notes)
```

Comment:
`code.notes` is now being iterated over twice! The condition `notes.isEmpty()` below can be replaced with simply `code.notes.empty()`.